tidy3d/components/simulation.py

```diff
@@ -242,6 +242,52 @@ def bloch_with_symmetry(cls, val, values):
                 )
         return val
 
+    # pylint: disable=too-many-locals
+    @pydantic.validator("boundary_spec", always=True)
+    def plane_wave_boundaries(cls, val, values):
+        """Error if there are plane wave sources incompatible with boundary conditions."""
+        boundaries = val.to_list
+        sources = values.get("sources")
+        size = values.get("size")
+        medium = values.get("medium")
+        for source in sources:
+            if not isinstance(source, PlaneWave):
+                continue
+            _, tan_dirs = cls.pop_axis([0, 1, 2], axis=source.injection_axis)
+            for tan_dir in tan_dirs:
+                boundary = boundaries[tan_dir]
+
+                # check the PML/absorber + angled plane wave case
+                num_pml = sum(isinstance(bnd, (PML, StablePML, Absorber)) for bnd in boundary)
+                if num_pml > 0 and (source.angle_theta != 0 or source.angle_phi != 0):
+                    raise SetupError(
+                        "Angled plane wave sources are not compatible with the absorbing or "
+                        f"perfectly matched boundary along dimension {tan_dir}. Either set the "
+                        "source angles to ``0``, or consider using Bloch boundaries."
+                    )
+
+                # check the Bloch boundary + angled plane wave case
+                num_bloch = sum(isinstance(bnd, BlochBoundary) for bnd in boundary)
+                if num_bloch > 0:
+                    # make a dummy Bloch boundary to check for correctness
+                    dummy_bnd = BlochBoundary.from_source(
+                        source=source, domain_size=size[tan_dir], axis=tan_dir, medium=medium
+                    )
+                    expected_bloch_vec = dummy_bnd.bloch_vec
+                    if boundary[0].bloch_vec != expected_bloch_vec:
+                        test_val = np.real(expected_bloch_vec - boundary[0].bloch_vec)
+                        if test_val.is_integer() and test_val != 0:
+                            log.warning(
+                                f"The Bloch vector along dimension {tan_dir} is not in the first "
+                                "Brillouin zone."
+                            )
+                        elif not test_val.is_integer():
+                            raise SetupError(
+                                f"The Bloch vector along dimension {tan_dir} is incorrectly set. "
```

Comment:
"The Bloch vector along dimension {tan_dir} is incorrectly set with respect to source at index {isrc}. Consider using ``BlochBoundary.from_source()``."